--- /mnt/cs519-sp25/linux-5.15.0/arch/x86/entry/syscalls/syscall_64.tbl.orig	2025-04-13 18:09:27.083067834 -0600
+++ /mnt/cs519-sp25/linux-5.15.0/arch/x86/entry/syscalls/syscall_64.tbl	2025-04-13 16:58:17.230261599 -0600
@@ -370,6 +370,8 @@
 446	common	landlock_restrict_self	sys_landlock_restrict_self
 447	common	memfd_secret		sys_memfd_secret
 448	common	process_mrelease	sys_process_mrelease
+449     common  enable_rb_extent        sys_enable_rb_extent
+450     common  disable_rb_extent       sys_disable_rb_extent
 
 #
 # Due to a historical design error, certain syscalls are numbered differently
